app/scripts/controllers/detect-tokens.js

```diff
@@ -0,0 +1,93 @@
+const Web3 = require('web3')
+const contracts = require('eth-contract-metadata')
+const { warn } = require('loglevel')
+const {
+    MAINNET,
+    } = require('./network/enums')
+// By default, poll every 3 minutes
+const DEFAULT_INTERVAL = 180 * 1000
+const ERC20_ABI = [{'constant': true, 'inputs': [{'name': '_owner', 'type': 'address'}], 'name': 'balanceOf', 'outputs': [{'name': 'balance', 'type': 'uint256'}], 'payable': false, 'type': 'function'}]
+
+/**
+ * A controller that polls for token exchange
+ * rates based on a user's current token list
+ */
+class DetectTokensController {
+  /**
+   * Creates a DetectTokensController
+   *
+   * @param {Object} [config] - Options to configure controller
+   */
+  constructor ({ interval = DEFAULT_INTERVAL, preferences, network } = {}) {
+    this.preferences = preferences
+    this.interval = interval
+    this.network = network
+  }
+
+   /**
+   * For each token in eth-contract-metada, find check selectedAddress balance.
+   *
+   */
+  async exploreNewTokens () {
```

Comment:
I just tried it out. Every three minutes may be fine for a background poll, but if we really want this feature, we should make it responsive.

I think we should also trigger this function from the `metamask-controller` when:
- `submitPassword` is [successfully called](https://github.com/MetaMask/metamask-extension/blob/develop/app/scripts/metamask-controller.js#L491).
- When a new account is selected (It will be something like `this.preferencesController.store.subscribe((state) => /* check for new value here... we should discuss the best way to do this... makes me wish we were using Ember Object... */

Reply:
It totally makes sense. 
Actually in line 80 of `detect-tokens.js` there is a selectedAddress update each time that the address is changed that could be used to trigger this function or also to restart the polling period, what do you think about that?

Reply:
Ok, I did it in a way that each time that the address is updated or there is a password submitted this function is going to be triggered and the polling is going to be initialized again. All from the `detect-tokens` controller to avoid dependency between those controllers.

Reply:
Also rather than using the `keyringController` as a way to detect a `submitPasword` event I used the same method already implemented to detect when the extension is active and therefore the user logged in.

Reply:
needs change on this last item